### ISSUE TITLE
tensor/generator input

### ISSUE THREAD

**Issue by ahundt** [titu1994/pyshac#3]
I was looking through the code the other day and some of it appears to assume the dataset can fit in memory as a numpy array. My dataset is 0.5 TB so unfortunately that won't work. Am I mistaken or will changes be required to account for this situation?

**Comment by titu1994** [comment]
Right currently it assumes the dataset path can be loaded in memory. 

The thing is, it uses XGBoost internally which cant handle (to my knowledge) 500 GB of data directly. 
Due to this, the best case would be to sort your dataset and pick only the top 100k or 1m rows of samples in your dataset, and write that to a csv file externally according to the specs of the dataset.

Also, with that much data, the classifiers should be able to do some pretty useful things I think.

**Comment by ahundt** [comment]
Sorry, I meant the application dataset (images, etc) rather than the hyperparameters dataset.

**Comment by titu1994** [comment]
Ah I mixed them up. 

No, there's no such restriction of the dataset having to be in memory. The evaluation function simply expects a final float value as the result of providing the hyper parameters. 

You could use TF to create the model, tf.Dataset API to load TFRecord dataset of any size, create the train graph and train a model for however many epochs you require and the final output would be the score on your validation accuracy (presumably the thing you wish to maximize). Only this is returned from each evaluation function.

The evaluation function is run on a different process (1 process for each available GPU), so you would need a powerful CPU / server to process the dataset of multiple GPUs in parallel, but that is left independent of the engine itself.

I have a micro example of this use case in the examples - [TF example](https://github.com/titu1994/pyshac/blob/master/examples/tensorflow/mini_conv.py). I used the tf.keras.datasets api to load datasets into numpy, but you could just as well use tf.Dataset with TFRecord to do the same, since tf.keras supports tf.Dataset api implicitly. 

At worst, you would need to change the model.fit() call to model.fit_generator()

**Comment by ahundt** [comment]
ah ok, thanks!

So predict(5) generates 5 lists of hyperparameter choices which pass the classifier requirements, or does it produce 5 classifiers like the comment seems to say?
https://github.com/titu1994/pyshac/blob/master/examples/tensorflow/mini_conv.py#L151

**Comment by titu1994** [comment]
Oh I should update that comment, it's quite ambiguous. 

Predict will return a list of ordered dictionaries, where each item i in the list can be used to construct a single model. (it is a list of sampled hyper parameters).

**Comment by ahundt** [comment]
Cool, that sounds good. 

Will they be encoded or already decoded from the search space?
Can we add these predictions to the dataset too?

In my own code I pretty much create an ever expanding "dataset" where all configurations & lengths of runs can be chopped down/skipped/incorporated as appropriate. To explain what I mean, imagine a search of the space containing 1 epoch runs, 40 epoch runs, and 100 epoch runs, you can use the 1st epoch of 100 epoch runs in your subset containing 1 epoch runs, the 40 can contain data from 40 and 100, but you'd have to skip scores of 1 epoch runs and 40 epoch runs when comparing among the 100 epoch runs.

**Comment by ahundt** [comment]
I'm hoping to release all my code publicly in the next couple weeks so I can be much more clear and specific in the near future.

**Comment by titu1994** [comment]
> Will they be encoded or already decoded from the search space?

They will be decoded values, and you can access them by the name of the parameter. So suppose you had a UniformContiniousHyperParameter named 'x' in the engine, then you can for each ordered dictionary you can simply write `sample['x']` to get the decoded value.

Same for discrete values, say you had a Hyper parameter 'y' and it can take the values [32, 64, 96, 128], then simply writing `sample['y']` will return one of the 4 values above for direct usage.

If you have a large list of parameters that you are going to map 1:1 with some function, then simply using `list(sample.values())` will give you a python list of the sampled values from each of the hyper parameters. 

> Can we add these predictions to the dataset too?

The dataset contains only the "value" of the sample space of that parameter. The configuration is setup independtly when you create the SHAC or TensorflowSHAC engine and pass the list of hyperparameters to it.

Your ever expanding dataset can be managed in two ways. 

1) You create an additional discrete parameter taking values [1, 40, 100] and augment your dataset with this information. In doing so, the cascade of classifiers will probably be biased towards the higher number of epochs as they often get better scores, so I wont recommend this approach.

2) Another way is, slice the data as provided, such they can be batched sequentially, and use `presort=False` flag inside `fit_dataset`. In doing so, each batch will take data as slices that you provided, inherently informing the model as to this hidden information. However, you need to be careful as to how you slice the data and how many batches you use.
